src/utils.c

```diff
@@ -110,30 +117,208 @@ SEXP colnamesInt(SEXP x, SEXP cols, SEXP check_dups) {
+  if (isLanguage(expr) && CAR(expr)==sym_colon) {  // 3:2, V3:V2, min(V3):min(V2)
+    //Rprintf("isLanguage(expr) && CAR(expr)==':'\n");
+    SEXP lhs = CADR(expr), rhs = CADDR(expr);
+    if (isSymbol(lhs) && isSymbol(rhs)) { // V3:V2
+      //Rprintf("isSymbol(colon_lhs) && isSymbol(colon_rhs)\n");
+      if (inverse)
+        error("column range selection by col1:col2 is not supported with combination of negation via '!' or '-'");
+      lhs = colnamesInt(x, ScalarString(PRINTNAME(lhs)), ScalarLogical(false), ScalarLogical(false)); // may raise error if lhs column does not exists
+      rhs = colnamesInt(x, ScalarString(PRINTNAME(rhs)), ScalarLogical(false), ScalarLogical(false));
+      if (!isInteger(lhs) || !isInteger(rhs) || length(lhs)!=1 || length(rhs)!=1 || LOGICAL(lhs)[0]==NA_LOGICAL || LOGICAL(rhs)[0]==NA_LOGICAL)
+        error(_("internal error: LHS and RHS of `:` call should be integer non-NA scalars already")); // # nocov
+      UNPROTECT(protecti);
+      return range_int(INTEGER(lhs)[0], INTEGER(rhs)[0]);
+    } else { // evaluates later on: 3:2, f(V3):f(V2)
+      //Rprintf("!isSymbol(colon_lhs) || !isSymbol(colon_rhs)\n");
+    }
+  }
+  SEXP cols = R_NilValue, value = R_NilValue;
+  if (isLanguage(expr) && CAR(expr)==sym_patterns) {
+    //Rprintf("isLanguage(expr) && CAR(expr)=='patterns'\n");
+    SEXP xnames = getAttrib(x, R_NamesSymbol);
+    if (isNull(xnames))
+      error(_("'x' argument data.table has no names"));
+    // .SDcols = Reduce(intersect, do_patterns(colsub, names_x))
+    SEXP sym_quote = install("quote");
+    SEXP sym_do_patterns = install("do_patterns");
+    SEXP lpat = PROTECT(eval(LCONS(sym_do_patterns, LCONS(LCONS(sym_quote, LCONS(expr, R_NilValue)), LCONS(xnames, R_NilValue))), rho)); protecti++;
+    SEXP sym_Reduce = install("Reduce");
+    SEXP sym_intersect = install("intersect");
+    cols = PROTECT(eval(LCONS(sym_Reduce, LCONS(sym_intersect, LCONS(lpat, R_NilValue))), rho)); protecti++;
+  }
+  // single symbol V2 might be also a function, to check that we need to evaluate, thus first we check if "V2" is existing column name, if not then we evaluate to see if it is a function.
+  // adding support of 'with' argument here may improve control, when with=FALSE then cols=eval(expr, rho) could be made straight away
+  if (isSymbol(expr)) { // V1, is.numeric
+    if (!isNull(cols))
+      error("internal error: cols should be still NULL at that point, did CAR(expr)=='patterns' branch kicked in? then isSymbol(expr) branch should not be reached"); // # nocov
+    SEXP xnames = getAttrib(x, R_NamesSymbol);
+    if (isNull(xnames))
+      error(_("'x' argument data.table has no names"));
+    cols = PROTECT(ScalarString(PRINTNAME(expr))); protecti++; // "V1", "is.numeric"
+    if (INTEGER(chmatch(cols, xnames, 0))[0]==0)
+      cols = R_NilValue;
+  }
+  // evaluate expr
+  if (isNull(cols)) {
+    value = PROTECT(eval(expr, rho)); protecti++;
```

Comment:
Note that ```value = PROTECT(eval(expr, rho)); protecti++;``` would allow ```j``` to accept a character vector without the ```..``` prefix which while I think is a good thing, would also be a change.
```
dt = as.data.table(lapply(1:5, c))
char_cols = c("V1", "V3")
.Call(data.table:::CexprCols, dt, quote(char_cols), environment()) 
[1] 1 3
```

Reply:
Handling of double dot prefix is currently not included in this PR, and it might eventually be better to polish out first the replacement for current mechanism before stepping into `..` prefix. Once we have this helper function done, it should be much easier to add an extra, recursive pre-processing for `..`, same like we do about `replace_dot_alias`.

Reply:
I am on Windows.

Reply:
do you have mingw 64bit or cygwin? it shoud mimic linux environment

Reply:
I will make a separate issue later for making ```cc()``` work for Windows user and then doing a PR for the developers markdown for what we  figure out. Thanks for bringing it up.

Reply:
best place for that will be https://github.com/Rdatatable/data.table/blob/master/.dev/README.md, to not increase the complexity I think the best would be to elaborate more on mingw usage for windows users rather then focusing on windows specific tools.